feat(keg): topic aliases (#5109)
Kong/developer.konghq.com: app/event-gateway/entities/virtual-cluster.md

```
client → policies → alias resolution → namespace → backend
```
{:.no-copy-code}

If the virtual cluster has a namespace, the alias must target a name that the namespace already exposes. For example, with `namespace.mode: hide_prefix` and `prefix: analytics_`, `topic: foo` resolves to the backend topic `analytics_foo`.

A backend topic that the namespace does not expose (it doesn't match the prefix and isn't listed in `additional_topics`) cannot be aliased directly. Expose it through the namespace first:

```yaml
namespace:
  mode: hide_prefix
  prefix: analytics_
  additional_topics:
    - type: exact_list
      list:
        - team-alpha-orders-v2
topic_aliases:
  - alias: orders
    topic: team-alpha-orders-v2 # valid because additional_topics exposes it
```

### Policy context

[Policies](/event-gateway/entities/policy/) attached to a virtual cluster see the alias name, not the backend topic name. In a CEL match expression, `topic.name` evaluates to the alias name. Policies that match on a backend topic name do not automatically apply to its aliases - match the alias name explicitly:

```yaml
condition: topic.name == "orders"
```

> For a full walkthrough, see [Configure topic aliases](/event-gateway/configure-topic-aliases/).

Reviewer: I think we need to add a point about not supporting consuming alias and non-alias topic with the same consumer group as the behaviour there is undefined

In this example, clients can produce and consume to `orders` and `clicks`. {{site.event_gateway_short}} resolves these to the backend topics `team-alpha-orders-v2` and `analytics-raw-clicks`.

### Supported operations

Aliases are a read-only abstraction over physical topics. Read and write operations are forwarded to the backend topic, but topic-modifying operations are rejected.

<!--vale off-->
{% table %}
columns:
  - title: Operation
    key: operation
  - title: Behavior
    key: behavior
rows:
  - operation: "Produce, fetch, list offsets, consumer group operations"
    behavior: "Allowed. Requests reference the alias name and are transparently resolved to the backend topic."
  - operation: "Metadata (`ListTopics`)"
    behavior: "Allowed. Both the alias and the original backend topic name appear in the response."
  - operation: "`CreateTopics`, `DeleteTopics`, `CreatePartitions`, `DeleteRecords`, `AlterPartitionReassignments`, `ElectLeaders`"
    behavior: "Rejected with `InvalidTopicException` when the request references an alias. This avoids a client unintentionally modifying a physical topic that other aliases or clients depend on."
{% endtable %}
<!--vale on-->

> ACLs are evaluated on the name the client uses, before alias resolution. An ACL on the backend topic does not automatically grant access to its aliases, and vice versa. Operators must configure ACLs for each alias name explicitly. With `acl_mode: enforce_on_gateway` (deny-by-default), a new alias with no matching ACL is blocked.

Reviewer: You should also mention:

1. that `topic.name` in the context is the name of the alias not the original name.
2. that `alias` applies post namespace so if a topic isn't in the namespace it should be added to the namespace first (this one is tricky so some examples and diagrams may help)

Author: done in next commit

]
```

> ## Topic aliases

Reviewer 2: ```suggestion
## Topic aliases {% new_in 1.2 %}
```

In this example, clients can produce and consume to `orders` and `clicks`. {{site.event_gateway_short}} resolves these to the backend topics `team-alpha-orders-v2` and `analytics-raw-clicks`.

### Supported operations

Aliases are a read-only abstraction over physical topics. Read and write operations are forwarded to the backend topic, but topic-modifying operations are rejected.

<!--vale off-->
{% table %}
columns:
  - title: Operation
    key: operation
  - title: Behavior
    key: behavior
rows:
  - operation: "Produce, fetch, list offsets, consumer group operations"
    behavior: "Allowed. Requests reference the alias name and are transparently resolved to the backend topic."
  - operation: "Metadata (`ListTopics`)"
    behavior: "Allowed. Both the alias and the original backend topic name appear in the response."
  - operation: "`CreateTopics`, `DeleteTopics`, `CreatePartitions`, `DeleteRecords`, `AlterPartitionReassignments`, `ElectLeaders`"
    behavior: "Rejected with `InvalidTopicException` when the request references an alias. This avoids a client unintentionally modifying a physical topic that other aliases or clients depend on."
{% endtable %}
<!--vale on-->

> ACLs are evaluated on the name the client uses, before alias resolution. An ACL on the backend topic does not automatically grant access to its aliases, and vice versa. Operators must configure ACLs for each alias name explicitly. With `acl_mode: enforce_on_gateway` (deny-by-default), a new alias with no matching ACL is blocked.

Reviewer 2: ```suggestion
[ACLs](/event-gateway/policies/acl/) are evaluated on the name the client uses, before alias resolution. An ACL on the backend topic does not automatically grant access to its aliases, and vice versa. Operators must configure ACLs for each alias name explicitly. With `acl_mode: enforce_on_gateway` (deny-by-default), a new alias with no matching ACL is blocked.
```

### Interaction with namespaces

The `topic` field in `topic_aliases` references namespace-visible names, not raw backend topic names. In the request path, aliases resolve *after* policies but *before* the [namespace](#namespaces) transformer:

```
client → policies → alias resolution → namespace → backend
```
{:.no-copy-code}

If the virtual cluster has a namespace, the alias must target a name that the namespace already exposes. For example, with `namespace.mode: hide_prefix` and `prefix: analytics_`, `topic: foo` resolves to the backend topic `analytics_foo`.

> A backend topic that the namespace does not expose (it doesn't match the prefix and isn't listed in `additional_topics`) cannot be aliased directly. Expose it through the namespace first:

Reviewer 2: ```suggestion
A backend topic that the namespace doesn't expose (not matching the prefix and not listed in `additional_topics`) can't be aliased directly. Expose it through the namespace first:
```

```yaml
namespace:
  mode: hide_prefix
  prefix: analytics_
  additional_topics:
    - type: exact_list
      list:
        - team-alpha-orders-v2
topic_aliases:
  - alias: orders
    topic: team-alpha-orders-v2 # valid because additional_topics exposes it
```

### Policy context

> [Policies](/event-gateway/entities/policy/) attached to a virtual cluster see the alias name, not the backend topic name. In a CEL match expression, `topic.name` evaluates to the alias name. Policies that match on a backend topic name do not automatically apply to its aliases - match the alias name explicitly:

Reviewer 2: ```suggestion
[Policies](/event-gateway/entities/policy/) attached to a virtual cluster see the alias name, not the backend topic name. In a [CEL match expression](/event-gateway/expressions/), `topic.name` evaluates to the alias name. Policies that match on a backend topic name don't automatically apply to its aliases. Match the alias name explicitly:
```

```yaml
condition: topic.name == "orders"
```

### Consumer groups

> Do not consume from both an alias and its backend topic with the same consumer group.

Reviewer 2: ```suggestion
Avoid consuming from both an alias and its backend topic with the same consumer group.
```

> {{site.event_gateway_short}} treats the alias and the backend topic as distinct topic names, but they share the same underlying partitions. Mixing them in one consumer group leaves offset commits and partition assignment undefined.

Reviewer 2: ```suggestion
{{site.event_gateway_short}} treats the alias and the backend topic as distinct topic names that share the same underlying partitions. Mixing them in one consumer group leaves offset commits and partition assignment undefined.
```
